### PR TITLE
Review and refactor src/crypto/bn.js

### LINKED ISSUE #24: Review and refactor src/crypto/bn.js
Make it comply with AirBnb style guide.

### PR DESCRIPTION
**Pull request by slashrsm**
Closes #24

### CODE REVIEW

File: src/crypto/bn.js [hunk 1]
```diff
@@ -1,130 +1,156 @@
...
   if (buf.length === 0) {
-    return BN.fromBuffer(new Buffer([0]));
+    return BN.fromBuffer(Buffer.alloc([0]));
   }
 
-  var endian = 'big';
+  let endian = 'big';
   if (opts) {
-    endian = opts.endian;
+    endian = opts;
```
Comment (slashrsm) [inline src/crypto/bn.js]: This is most likely wrong. I assume that the intention was to use the [destructing assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment).

File: src/crypto/bn.js [hunk 2]
```diff
@@ -1,130 +1,156 @@
...
-    if (buf[0] & 0x80) {
-      buf = Buffer.concat([new Buffer([0x80]), buf]);
+    if (buf[0] && 0x80) {
+      buf = Buffer.concat([Buffer.alloc([0x80]), buf]);
     } else {
-      buf[0] = buf[0] | 0x80;
+      const tmpvar = buf[0];
+      /*eslint-disable */
+      buf[0] = tmpvar | 0x80;
+      /* eslint-enable */
```
Comment (slashrsm) [inline src/crypto/bn.js]: "no-bitwise" rule is disabled in the .eslintrc.json so this shouldn't be needed.

File: src/crypto/bn.js [hunk 3]
```diff
@@ -1,130 +1,156 @@
...
     }
   } else {
     buf = this.toBuffer();
+    /* eslint no-bitwise: ["error", { "allow": ["&"] }] */
```
Comment (slashrsm) [inline src/crypto/bn.js]: Should not be needed either.

File: src/crypto/bn.js [hunk 4]
```diff
@@ -1,130 +1,156 @@
...
  * (a buffer where the most significant bit represents the sign (0 = positive, -1 = negative))
  */
-BN.fromSM = function(buf, opts) {
-  var ret;
+function fromSM(buf, opts) {
+  let ret;
+  let localbuf;
   if (buf.length === 0) {
-    return BN.fromBuffer(new Buffer([0]));
+    return BN.fromBuffer(Buffer.alloc([0]));
```
Comment (paranojik) [inline src/crypto/bn.js]: You should probably use Buffer.from() here. See https://nodejs.org/api/buffer.html#buffer_class_method_buffer_from_buffer.

File: src/crypto/bn.js [hunk 5]
```diff
@@ -1,130 +1,156 @@
...
+function toBuffer(opts) {
+  let buf;
+  let hex;
   if (opts && opts.size) {
     hex = this.toString(16, 2);
-    var natlen = hex.length / 2;
+    const natlen = hex.length / 2;
+    /* eslint-disable */
     buf = new Buffer(hex, 'hex');
+      /* eslint-enable */
```
Comment (paranojik) [inline src/crypto/bn.js]: Again. User Buffer.from and remove eslint-disable.

File: src/crypto/bn.js [hunk 6]
```diff
@@ -1,130 +1,156 @@
...
-    } else if (natlen > opts.size) {
+    if (natlen > opts.size) {
       buf = BN.trim(buf, natlen);
     } else if (natlen < opts.size) {
       buf = BN.pad(buf, natlen, opts.size);
     }
   } else {
     hex = this.toString(16, 2);
-    buf = new Buffer(hex, 'hex');
+    buf = Buffer.alloc(hex, 'hex');
```
Comment (paranojik) [inline src/crypto/bn.js]: Buffer.from()

File: src/crypto/bn.js [hunk 7]
```diff
@@ -1,130 +1,156 @@
...
-BN.prototype.toSMBigEndian = function() {
-  var buf;
+function toSMBigEndian() {
+  let buf;
   if (this.cmp(BN.Zero) === -1) {
     buf = this.neg().toBuffer();
-    if (buf[0] & 0x80) {
-      buf = Buffer.concat([new Buffer([0x80]), buf]);
+    if (buf[0] && 0x80) {
+      buf = Buffer.concat([Buffer.alloc([0x80]), buf]);
```
Comment (paranojik) [inline src/crypto/bn.js]: Buffer.from()

File: src/crypto/bn.js [hunk 8]
```diff
@@ -1,130 +1,156 @@
...
+      /*eslint-disable */
+      buf[0] = tmpvar | 0x80;
+      /* eslint-enable */
     }
   } else {
     buf = this.toBuffer();
+    /* eslint no-bitwise: ["error", { "allow": ["&"] }] */
     if (buf[0] & 0x80) {
-      buf = Buffer.concat([new Buffer([0x00]), buf]);
+      buf = Buffer.concat([Buffer.alloc([0x00]), buf]);
```
Comment (paranojik) [inline src/crypto/bn.js]: ... same here...

File: src/crypto/bn.js [hunk 9]
```diff
@@ -1,130 +1,156 @@
...
     }
   }
 
-  if (buf.length === 1 & buf[0] === 0) {
-    buf = new Buffer([]);
+  if (buf.length === 1 && buf[0] === 0) {
+    buf = Buffer.alloc([]);
```
Comment (paranojik) [inline src/crypto/bn.js]: ...same here...

File: src/crypto/bn.js [hunk 10]
```diff
@@ -1,130 +1,156 @@
-'use strict';
 
-var BN = require('bn.js');
-var $ = require('../util/preconditions');
-var _ = require('lodash');
```
[the remark this answers is not in the document]
Comment (ClemensLey) [reply src/crypto/bn.js]: Yup!

**Pull request closed without merging**